### PR TITLE
fix(storage): update seal_epoch synchronously

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Before this PR, seal_epoch updating is handled by HummockEventHandler [asynchronously](https://github.com/risingwavelabs/risingwave/blob/12127933b945ba414e5a7329e04139c65f09f9c7/src/stream/src/task/barrier_manager/managed_state.rs#L133), which cannot ensure HummockEventHandler::seal_epoch is updated [before notifying barrier finishing ](https://github.com/risingwavelabs/risingwave/blob/12127933b945ba414e5a7329e04139c65f09f9c7/src/stream/src/task/barrier_manager/managed_state.rs#L144). It can lead to [current epoch can't read, because the epoch in storage is not updated](https://github.com/risingwavelabs/risingwave/issues/7169).

This PR fixes it by updating HummockStorage::seal_epoch synchronously, which shares the same underlying atomic var of HummockEventHandler::seal_epoch.

## Checklist

- [x] I have written necessary rustdoc comments
~~- [ ] I have added necessary unit tests and integration tests~~
~~- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)~~

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
fix https://github.com/risingwavelabs/risingwave/issues/7169

### CODE REVIEW

File: src/storage/src/hummock/state_store.rs [hunk 1]
```diff
@@ -268,6 +268,9 @@ impl StateStore for HummockStorage {
             warn!("sealing invalid epoch");
             return;
         }
+        // Update `seal_epoch` synchronously,
+        // as `HummockEvent::SealEpoch` is handled asynchronously.
+        self.seal_epoch.store(epoch, MemOrdering::SeqCst);
```
Comment (wenym1) [inline src/storage/src/hummock/state_store.rs:273]: Before this PR, the seal epoch should be updated in somewhere else. Should we remove updating the seal epoch there?

Comment (hzxa21) [reply src/storage/src/hummock/state_store.rs:273]: > Before this PR, the seal epoch should be updated in somewhere else. Should we remove updating the seal epoch there?

+1. I think we don't need to update seal epoch when handling `HummockEvent::SealEpoch` in `HummockEventHandler`.

Comment (zwang28) [reply src/storage/src/hummock/state_store.rs:273]: Fixed.

**Comment by wenym1** [comment]
If this PR is intended to fix the panic caused by `try_wait_epoch` on uncommitted epoch, I would suggest we modify the logic of `try_wait_epoch` to do a really wait for the `sealed_epoch` to bump up asynchronously. In the future there may not be such global `seal_epoch` method, but instead seal each local state store, and then we may need to wait on the `sealed_epoch` of each local state store instance to bump up.

**Review by wenym1 — APPROVED** [review]
Anyway, the current fix LGTM. We may introduce the real wait in future PR.

**Comment by zwang28** [comment]
> I would suggest we modify the logic of try_wait_epoch to do a really wait for the sealed_epoch to bump up asynchronously

Agree. 
I'll merge this quick fix first.

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**